nomad/state/state_store.go

```diff
@@ -1814,6 +1859,260 @@ func (s *StateStore) VaultAccessorsByNode(ws memdb.WatchSet, nodeID string) ([]*
+			continue
+		}
+
+		// Check that the canary is part of a group being promoted
+		inGroup, ok := req.Groups[alloc.TaskGroup]
+		if !req.All && (!ok || !inGroup) {
+			continue
+		}
+
+		// Ensure the canaries are healthy
+		if !alloc.DeploymentStatus.IsHealthy() {
+			multierror.Append(&unhealthyErr, fmt.Errorf("Canary allocation %q for group %q is not healthy", alloc.ID, alloc.TaskGroup))
+			continue
+		}
+
+		canaries = append(canaries, alloc)
+	}
+
+	if err := unhealthyErr.ErrorOrNil(); err != nil {
+		return err
+	}
+
+	// Mark canaries as promoted and upsert them
+	canariesCopy := make([]*structs.Allocation, 0, len(canaries))
+	for _, c := range canaries {
+		copy := c.Copy()
+		copy.DeploymentStatus.Promoted = true
+		copy.DeploymentStatus.ModifyIndex = index
+		canariesCopy = append(canariesCopy, copy)
+	}
+	if err := s.upsertAllocsImpl(index, canariesCopy, txn); err != nil {
+		return err
+	}
+
+	// Update deployment
+	copy := deployment.Copy()
+	copy.ModifyIndex = index
+	for tg, status := range copy.TaskGroups {
+		inGroup, ok := req.Groups[tg]
+		if !req.All && (!ok || !inGroup) {
+			continue
+		}
+
+		status.Promoted = true
+	}
+
+	// Insert the deployment
+	if err := txn.Insert("deployment", copy); err != nil {
+		return err
+	}
+
+	// Update the index
+	if err := txn.Insert("index", &IndexEntry{"deployment", index}); err != nil {
+		return fmt.Errorf("index update failed: %v", err)
+	}
+
+	// Upsert the optional eval
+	if req.Eval != nil {
+		if err := s.nestedUpsertEval(txn, index, req.Eval); err != nil {
+			return err
+		}
+	}
+
+	txn.Commit()
+	return nil
+}
+
+// UpsertDeploymentAllocHealth is used to update the health of allocations as
+// part of the deployment and potentially make a evaluation
+func (s *StateStore) UpsertDeploymentAllocHealth(index uint64, req *structs.ApplyDeploymentAllocHealthRequest) error {
+	txn := s.db.Txn(true)
+	defer txn.Abort()
+
+	// Retrieve deployment and ensure it is not terminal and is active
+	ws := memdb.NewWatchSet()
+	deployment, err := s.deploymentByIDImpl(ws, req.DeploymentID, txn)
+	if err != nil {
+		return err
+	} else if deployment == nil {
+		return fmt.Errorf("Deployment ID %q couldn't be updated as it does not exist", req.DeploymentID)
+	} else if !deployment.Active() {
+		return fmt.Errorf("Deployment %q has terminal status %q:", deployment.ID, deployment.Status)
+	}
+
+	// Update the health status of each allocation
+	if total := len(req.HealthyAllocationIDs) + len(req.UnhealthyAllocationIDs); total != 0 {
+		setAllocHealth := func(id string, healthy bool) error {
+			existing, err := txn.First("allocs", "id", id)
+			if err != nil {
+				return fmt.Errorf("alloc %q lookup failed: %v", id, err)
+			}
+			if existing == nil {
+				return fmt.Errorf("unknown alloc %q", id)
+			}
+
+			old := existing.(*structs.Allocation)
+			if old.DeploymentID != req.DeploymentID {
+				return fmt.Errorf("alloc %q is not part of deployment %q", id, req.DeploymentID)
+			}
+
```

Comment:
Make sure that the alloc is actually healthy when setting it to healthy, otherwise error